{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "played-rover",
   "metadata": {},
   "source": [
    "### Merge data from different lanes and remove doublet (patient 5,6,7,8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "decent-ceiling",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: If you miss a compact list, please try `print_header`!\n",
      "The `sinfo` package has changed name and is now called `session_info` to become more discoverable and self-explanatory. The `sinfo` PyPI package will be kept around to avoid breaking old installs and you can downgrade to 0.3.2 if you want to use it without seeing this message. For the latest features and bug fixes, please install `session_info` instead. The usage and defaults also changed slightly, so please review the latest README at https://gitlab.com/joelostblom/session_info.\n",
      "-----\n",
      "anndata     0.7.6\n",
      "scanpy      1.8.1\n",
      "sinfo       0.3.4\n",
      "-----\n",
      "PIL                 8.1.2\n",
      "annoy               NA\n",
      "backcall            0.2.0\n",
      "bbknn               NA\n",
      "cffi                1.15.0\n",
      "cloudpickle         2.1.0\n",
      "colorama            0.4.4\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "dask                2022.02.0\n",
      "dateutil            2.8.1\n",
      "decorator           4.4.2\n",
      "fsspec              2021.10.0\n",
      "google              NA\n",
      "h5py                2.10.0\n",
      "igraph              0.8.0\n",
      "ipykernel           5.5.0\n",
      "ipython_genutils    0.2.0\n",
      "jedi                0.18.0\n",
      "jinja2              3.1.2\n",
      "joblib              1.0.1\n",
      "kiwisolver          1.3.1\n",
      "leidenalg           0.8.3\n",
      "llvmlite            0.36.0\n",
      "louvain             0.6.1\n",
      "markupsafe          2.1.1\n",
      "matplotlib          3.4.0\n",
      "mpl_toolkits        NA\n",
      "natsort             7.1.1\n",
      "numba               0.53.1\n",
      "numexpr             2.7.3\n",
      "numpy               1.21.2\n",
      "packaging           20.9\n",
      "pandas              1.2.4\n",
      "parso               0.8.1\n",
      "pexpect             4.8.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "prompt_toolkit      3.0.18\n",
      "psutil              5.8.0\n",
      "ptyprocess          0.7.0\n",
      "pycparser           2.21\n",
      "pygments            2.8.1\n",
      "pyparsing           2.4.7\n",
      "pytoml              NA\n",
      "pytz                2022.1\n",
      "scipy               1.6.3\n",
      "scrublet            NA\n",
      "setuptools_scm      NA\n",
      "six                 1.15.0\n",
      "sklearn             0.24.1\n",
      "statsmodels         0.11.1\n",
      "storemagic          NA\n",
      "tables              3.6.1\n",
      "tblib               1.7.0\n",
      "texttable           1.6.3\n",
      "tlz                 0.12.0\n",
      "toolz               0.12.0\n",
      "tornado             6.1\n",
      "traitlets           5.0.5\n",
      "typing_extensions   NA\n",
      "umap                0.3.10\n",
      "wcwidth             0.2.5\n",
      "yaml                5.4.1\n",
      "zipp                NA\n",
      "zmq                 22.0.3\n",
      "-----\n",
      "IPython             7.21.0\n",
      "jupyter_client      6.1.12\n",
      "jupyter_core        4.7.1\n",
      "notebook            6.4.4\n",
      "-----\n",
      "Python 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37) [GCC 9.3.0]\n",
      "Linux-4.15.0-158-generic-x86_64-with-debian-bullseye-sid\n",
      "60 logical CPU cores, x86_64\n",
      "-----\n",
      "Session information updated at 2022-08-31 10:42\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scrublet as scr\n",
    "import sys\n",
    "import bbknn\n",
    "from statsmodels import robust\n",
    "import matplotlib.pyplot as plt\n",
    "import os.path\n",
    "\n",
    "sc.settings.verbosity = 3  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "#sc.settings.set_figure_params(dpi=80, color_map='viridis')\n",
    "sc.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "dress-modern",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Package                       Version\n",
      "----------------------------- -------------------\n",
      "absl-py                       0.14.1\n",
      "adjustText                    0.8b2\n",
      "aiohttp                       3.7.4.post0\n",
      "airr                          1.3.1\n",
      "anndata                       0.7.6\n",
      "anndata2ri                    1.0.6\n",
      "annoy                         1.17.0\n",
      "argon2-cffi                   21.1.0\n",
      "async-timeout                 3.0.1\n",
      "attrs                         21.2.0\n",
      "backcall                      0.2.0\n",
      "backports.functools-lru-cache 1.6.1\n",
      "backports.zoneinfo            0.2.1\n",
      "bbknn                         1.4.1\n",
      "bleach                        4.1.0\n",
      "bokeh                         2.4.3\n",
      "cached-property               1.5.2\n",
      "cachetools                    4.2.4\n",
      "celltypist                    1.0.0\n",
      "certifi                       2020.12.5\n",
      "cffi                          1.15.0\n",
      "chardet                       4.0.0\n",
      "charset-normalizer            2.0.6\n",
      "click                         8.0.1\n",
      "cloudpickle                   2.1.0\n",
      "cmake                         3.18.4.post1\n",
      "colorama                      0.4.4\n",
      "commonmark                    0.9.1\n",
      "cycler                        0.10.0\n",
      "Cython                        0.29.22\n",
      "dask                          2022.2.0\n",
      "decorator                     4.4.2\n",
      "defusedxml                    0.7.1\n",
      "distributed                   2022.2.0\n",
      "entrypoints                   0.3\n",
      "et-xmlfile                    1.1.0\n",
      "fa2                           0.3.5\n",
      "fbpca                         1.0\n",
      "fcsparser                     0.2.1\n",
      "filelock                      3.3.0\n",
      "fsspec                        2021.10.0\n",
      "future                        0.18.2\n",
      "gdown                         4.0.2\n",
      "geosketch                     1.2\n",
      "get-version                   2.1\n",
      "google-auth                   1.35.0\n",
      "google-auth-oauthlib          0.4.6\n",
      "grpcio                        1.41.0\n",
      "h5py                          2.10.0\n",
      "harmonypy                     0.0.5\n",
      "harmonyTS                     0.1.4\n",
      "HeapDict                      1.0.1\n",
      "idna                          3.2\n",
      "imageio                       2.9.0\n",
      "importlib-metadata            1.7.0\n",
      "ipykernel                     5.5.0\n",
      "ipython                       7.21.0\n",
      "ipython-genutils              0.2.0\n",
      "ipywidgets                    7.6.5\n",
      "jedi                          0.18.0\n",
      "Jinja2                        3.1.2\n",
      "joblib                        1.0.1\n",
      "jsonschema                    4.0.1\n",
      "jupyter-client                6.1.12\n",
      "jupyter-core                  4.7.1\n",
      "jupyterlab-pygments           0.1.2\n",
      "jupyterlab-widgets            1.0.2\n",
      "kiwisolver                    1.3.1\n",
      "legacy-api-wrap               1.2\n",
      "leidenalg                     0.8.3\n",
      "llvmlite                      0.36.0\n",
      "locket                        1.0.0\n",
      "loompy                        3.0.6\n",
      "louvain                       0.6.1\n",
      "Markdown                      3.3.4\n",
      "MarkupSafe                    2.1.1\n",
      "matplotlib                    3.4.0\n",
      "memory-profiler               0.58.0\n",
      "mistune                       0.8.4\n",
      "msgpack                       1.0.4\n",
      "MulticoreTSNE                 0.1\n",
      "multidict                     5.2.0\n",
      "natsort                       7.1.1\n",
      "nbclient                      0.5.4\n",
      "nbconvert                     6.2.0\n",
      "nbformat                      5.1.3\n",
      "nest-asyncio                  1.5.1\n",
      "networkx                      2.5\n",
      "notebook                      6.4.4\n",
      "numba                         0.53.1\n",
      "numexpr                       2.7.3\n",
      "numpy                         1.21.2\n",
      "numpy-groupies                0.9.13\n",
      "oauthlib                      3.1.1\n",
      "openpyxl                      3.0.9\n",
      "opt-einsum                    3.3.0\n",
      "packaging                     20.9\n",
      "palantir                      1.0.0\n",
      "pandas                        1.2.4\n",
      "pandocfilters                 1.5.0\n",
      "parasail                      1.2.4\n",
      "parso                         0.8.1\n",
      "partd                         1.2.0\n",
      "patsy                         0.5.1\n",
      "pexpect                       4.8.0\n",
      "PhenoGraph                    1.5.7\n",
      "pickleshare                   0.7.5\n",
      "Pillow                        8.1.2\n",
      "pip                           21.0.1\n",
      "prometheus-client             0.11.0\n",
      "prompt-toolkit                3.0.18\n",
      "protobuf                      3.18.1\n",
      "psutil                        5.8.0\n",
      "ptyprocess                    0.7.0\n",
      "pyasn1                        0.4.8\n",
      "pyasn1-modules                0.2.8\n",
      "pycparser                     2.21\n",
      "pyDeprecate                   0.3.0\n",
      "pydot                         1.4.2\n",
      "Pygments                      2.8.1\n",
      "pynndescent                   0.5.2\n",
      "pyparsing                     2.4.7\n",
      "pyro-api                      0.1.2\n",
      "pyro-ppl                      1.7.0\n",
      "pyrsistent                    0.18.0\n",
      "PySocks                       1.7.1\n",
      "python-dateutil               2.8.1\n",
      "python-igraph                 0.8.0\n",
      "python-Levenshtein            0.12.2\n",
      "pytoml                        0.1.21\n",
      "pytorch-lightning             1.3.8\n",
      "pytz                          2022.1\n",
      "pytz-deprecation-shim         0.1.0.post0\n",
      "PyWavelets                    1.1.1\n",
      "PyYAML                        5.4.1\n",
      "pyzmq                         22.0.3\n",
      "requests                      2.26.0\n",
      "requests-oauthlib             1.3.0\n",
      "rich                          10.12.0\n",
      "rpy2                          3.4.2\n",
      "rsa                           4.7.2\n",
      "scanpy                        1.8.1\n",
      "scArches                      0.4.0\n",
      "scIB                          0.1.1\n",
      "scikit-image                  0.18.1\n",
      "scikit-learn                  0.24.1\n",
      "scikit-misc                   0.1.4\n",
      "scipy                         1.6.3\n",
      "scirpy                        0.7.0\n",
      "scrublet                      0.2.3\n",
      "sctk                          0.0.1\n",
      "scvi-tools                    0.13.0\n",
      "seaborn                       0.11.1\n",
      "Send2Trash                    1.8.0\n",
      "setuptools                    49.6.0.post20210108\n",
      "setuptools-scm                6.0.1\n",
      "sinfo                         0.3.4\n",
      "six                           1.15.0\n",
      "sklearn                       0.0\n",
      "sortedcontainers              2.4.0\n",
      "squarify                      0.4.3\n",
      "statsmodels                   0.11.1\n",
      "stdlib-list                   0.8.0\n",
      "tables                        3.6.1\n",
      "tblib                         1.7.0\n",
      "tensorboard                   2.6.0\n",
      "tensorboard-data-server       0.6.1\n",
      "tensorboard-plugin-wit        1.8.0\n",
      "terminado                     0.12.1\n",
      "testpath                      0.5.0\n",
      "texttable                     1.6.3\n",
      "threadpoolctl                 2.1.0\n",
      "tifffile                      2021.4.8\n",
      "toolz                         0.12.0\n",
      "torch                         1.9.1\n",
      "torchmetrics                  0.5.1\n",
      "tornado                       6.1\n",
      "tqdm                          4.59.0\n",
      "traitlets                     5.0.5\n",
      "typing-extensions             4.3.0\n",
      "tzdata                        2022.1\n",
      "tzlocal                       4.2\n",
      "umap-learn                    0.3.10\n",
      "urllib3                       1.26.7\n",
      "wcwidth                       0.2.5\n",
      "webencodings                  0.5.1\n",
      "Werkzeug                      2.0.2\n",
      "wheel                         0.36.2\n",
      "widgetsnbextension            3.5.1\n",
      "xlrd                          1.2.0\n",
      "yamlordereddictloader         0.4.0\n",
      "yarl                          1.7.0\n",
      "zict                          2.2.0\n",
      "zipp                          3.4.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "awful-firewall",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CTCL5_Derm_45P_8N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCL5_Derm_45P_8Pr_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTCL5_Epi_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTCL5_Epi_45P_8N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CTCL5_Epi_45P_8Pr_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CTCL6_Derm45N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CTCL6_Derm45P_8N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CTCL6_Derm45P_8Pr_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CTCL6_Epi45N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CTCL6_Epi45P_8N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CTCL6_Epi45P_8Pr_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CTCL7_Derm45N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CTCL7_Derm45P_8N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CTCL7_Derm45P_8Pr_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CTCL7_Epi45N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CTCL7_Epi45P_8N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CTCL7_Epi45P_8Pr_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CTCL8_Derm45N_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CTCL8_Derm45P_8N_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CTCL8_Derm45P_8Pr_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CTCL8_Epi45N_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_id donor_id     Tissue  id\n",
       "0       CTCL5_Derm_45N_G    CTCL5     Dermis  18\n",
       "1    CTCL5_Derm_45P_8N_G    CTCL5     Dermis  19\n",
       "2   CTCL5_Derm_45P_8Pr_G    CTCL5     Dermis  20\n",
       "3        CTCL5_Epi_45N_G    CTCL5  Epidermis  21\n",
       "4     CTCL5_Epi_45P_8N_G    CTCL5  Epidermis  22\n",
       "5    CTCL5_Epi_45P_8Pr_G    CTCL5  Epidermis  23\n",
       "6        CTCL6_Derm45N_G    CTCL6     Dermis  24\n",
       "7     CTCL6_Derm45P_8N_G    CTCL6     Dermis  25\n",
       "8    CTCL6_Derm45P_8Pr_G    CTCL6     Dermis  26\n",
       "9         CTCL6_Epi45N_G    CTCL6  Epidermis  27\n",
       "10     CTCL6_Epi45P_8N_G    CTCL6  Epidermis  28\n",
       "11    CTCL6_Epi45P_8Pr_G    CTCL6  Epidermis  29\n",
       "12       CTCL7_Derm45N_G    CTCL7     Dermis  30\n",
       "13    CTCL7_Derm45P_8N_G    CTCL7     Dermis  31\n",
       "14   CTCL7_Derm45P_8Pr_G    CTCL7     Dermis  32\n",
       "15        CTCL7_Epi45N_G    CTCL7  Epidermis  33\n",
       "16     CTCL7_Epi45P_8N_G    CTCL7  Epidermis  34\n",
       "17    CTCL7_Epi45P_8Pr_G    CTCL7  Epidermis  35\n",
       "18       CTCL8_Derm45N_G    CTCL8     Dermis  36\n",
       "19    CTCL8_Derm45P_8N_G    CTCL8     Dermis  37\n",
       "20   CTCL8_Derm45P_8Pr_G    CTCL8     Dermis  38\n",
       "21        CTCL8_Epi45N_G    CTCL8  Epidermis  39\n",
       "22   CTCL8_Epi45P_POOL_G    CTCL8  Epidermis  40"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"/lustre/scratch117/cellgen/team205/rl20/CTCL/Cellranger_out\"\n",
    "manifest = pd.read_csv(os.path.join(path, 'sample_manifest.csv'))\n",
    "manifest\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "hidden-grave",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = manifest['sample_id'].tolist()\n",
    "filenames = [os.path.join(path, n, 'filtered_feature_bc_matrix') for n in names]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "romance-banner",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID = manifest['id'].tolist()\n",
    "ID = list(map(str, ID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "statewide-sentence",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n",
      "... writing an h5ad cache file to speedup reading next time\n"
     ]
    }
   ],
   "source": [
    "adatas = [sc.read_10x_mtx(filename, cache=True) for filename in filenames]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "federal-calendar",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = adatas[0].concatenate(adatas[1:], join='inner', batch_key='ID', batch_categories=ID, index_unique='-')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "cosmetic-composer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CTCL5_Derm_45P_8N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTCL5_Derm_45P_8Pr_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTCL5_Epi_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTCL5_Epi_45P_8N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CTCL5_Epi_45P_8Pr_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CTCL6_Derm45N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CTCL6_Derm45P_8N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CTCL6_Derm45P_8Pr_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CTCL6_Epi45N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CTCL6_Epi45P_8N_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CTCL6_Epi45P_8Pr_G</td>\n",
       "      <td>CTCL6</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CTCL7_Derm45N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CTCL7_Derm45P_8N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CTCL7_Derm45P_8Pr_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CTCL7_Epi45N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CTCL7_Epi45P_8N_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CTCL7_Epi45P_8Pr_G</td>\n",
       "      <td>CTCL7</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>CTCL8_Derm45N_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>CTCL8_Derm45P_8N_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CTCL8_Derm45P_8Pr_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>CTCL8_Epi45N_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_id donor_id     Tissue  id\n",
       "0       CTCL5_Derm_45N_G    CTCL5     Dermis  18\n",
       "1    CTCL5_Derm_45P_8N_G    CTCL5     Dermis  19\n",
       "2   CTCL5_Derm_45P_8Pr_G    CTCL5     Dermis  20\n",
       "3        CTCL5_Epi_45N_G    CTCL5  Epidermis  21\n",
       "4     CTCL5_Epi_45P_8N_G    CTCL5  Epidermis  22\n",
       "5    CTCL5_Epi_45P_8Pr_G    CTCL5  Epidermis  23\n",
       "6        CTCL6_Derm45N_G    CTCL6     Dermis  24\n",
       "7     CTCL6_Derm45P_8N_G    CTCL6     Dermis  25\n",
       "8    CTCL6_Derm45P_8Pr_G    CTCL6     Dermis  26\n",
       "9         CTCL6_Epi45N_G    CTCL6  Epidermis  27\n",
       "10     CTCL6_Epi45P_8N_G    CTCL6  Epidermis  28\n",
       "11    CTCL6_Epi45P_8Pr_G    CTCL6  Epidermis  29\n",
       "12       CTCL7_Derm45N_G    CTCL7     Dermis  30\n",
       "13    CTCL7_Derm45P_8N_G    CTCL7     Dermis  31\n",
       "14   CTCL7_Derm45P_8Pr_G    CTCL7     Dermis  32\n",
       "15        CTCL7_Epi45N_G    CTCL7  Epidermis  33\n",
       "16     CTCL7_Epi45P_8N_G    CTCL7  Epidermis  34\n",
       "17    CTCL7_Epi45P_8Pr_G    CTCL7  Epidermis  35\n",
       "18       CTCL8_Derm45N_G    CTCL8     Dermis  36\n",
       "19    CTCL8_Derm45P_8N_G    CTCL8     Dermis  37\n",
       "20   CTCL8_Derm45P_8Pr_G    CTCL8     Dermis  38\n",
       "21        CTCL8_Epi45N_G    CTCL8  Epidermis  39\n",
       "22   CTCL8_Epi45P_POOL_G    CTCL8  Epidermis  40"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "usual-egypt",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_id\n",
      "donor_id\n",
      "Tissue\n"
     ]
    }
   ],
   "source": [
    "#%% Metadata\n",
    "for metadata in manifest.columns[0:-1]: # sample_id and vdj not needed\n",
    "    print(metadata)\n",
    "    adata.obs[metadata] = adata.obs['ID']\n",
    "    replacement = manifest[metadata].tolist()\n",
    "    adata.obs[metadata].replace(to_replace=ID, value=replacement, inplace=True)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "large-section",
   "metadata": {},
   "outputs": [],
   "source": [
    "del(adata.obs['ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "dietary-repair",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTGATAA-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCATCCA-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAGCCGTC-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACCGTTGG-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTCCCTACT-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAACGCTA-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCACGCGGT-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCAGATCA-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGGTCTAA-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTGTATC-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>206362 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 sample_id donor_id     Tissue\n",
       "AAACCTGAGCTGATAA-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis\n",
       "AAACCTGAGTCATCCA-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis\n",
       "AAACCTGCAAGCCGTC-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis\n",
       "AAACCTGCACCGTTGG-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis\n",
       "AAACCTGGTCCCTACT-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis\n",
       "...                                    ...      ...        ...\n",
       "TTTGTCATCAACGCTA-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis\n",
       "TTTGTCATCACGCGGT-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis\n",
       "TTTGTCATCCAGATCA-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis\n",
       "TTTGTCATCGGTCTAA-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis\n",
       "TTTGTCATCTTGTATC-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis\n",
       "\n",
       "[206362 rows x 3 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "global-chuck",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sample_id' as categorical\n",
      "... storing 'donor_id' as categorical\n",
      "... storing 'Tissue' as categorical\n",
      "... storing 'feature_types' as categorical\n"
     ]
    }
   ],
   "source": [
    "save_file = '/lustre/scratch117/cellgen/team205/rl20/CTCL/object-new/CTCL5678_raw.h5ad'\n",
    "adata.write(save_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "temporal-buffer",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = '/lustre/scratch117/cellgen/team205/rl20/CTCL/object-new/CTCL5678_raw.h5ad'\n",
    "adata = sc.read_h5ad(input_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "curious-persian",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.3</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.2</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC233755.1</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC240274.1</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC213203.1</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM231C</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>Gene Expression</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ids    feature_types\n",
       "MIR1302-2HG  ENSG00000243485  Gene Expression\n",
       "FAM138A      ENSG00000237613  Gene Expression\n",
       "OR4F5        ENSG00000186092  Gene Expression\n",
       "AL627309.1   ENSG00000238009  Gene Expression\n",
       "AL627309.3   ENSG00000239945  Gene Expression\n",
       "...                      ...              ...\n",
       "AC233755.2   ENSG00000277856  Gene Expression\n",
       "AC233755.1   ENSG00000275063  Gene Expression\n",
       "AC240274.1   ENSG00000271254  Gene Expression\n",
       "AC213203.1   ENSG00000277475  Gene Expression\n",
       "FAM231C      ENSG00000268674  Gene Expression\n",
       "\n",
       "[33538 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "southwest-bishop",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238009</th>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AL627309.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000239945</th>\n",
       "      <td>ENSG00000239945</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AL627309.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277856</th>\n",
       "      <td>ENSG00000277856</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC233755.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000275063</th>\n",
       "      <td>ENSG00000275063</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC233755.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000271254</th>\n",
       "      <td>ENSG00000271254</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC240274.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000277475</th>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>AC213203.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268674</th>\n",
       "      <td>ENSG00000268674</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>FAM231C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33538 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_ids    feature_types         gene\n",
       "ENSG00000243485  ENSG00000243485  Gene Expression  MIR1302-2HG\n",
       "ENSG00000237613  ENSG00000237613  Gene Expression      FAM138A\n",
       "ENSG00000186092  ENSG00000186092  Gene Expression        OR4F5\n",
       "ENSG00000238009  ENSG00000238009  Gene Expression   AL627309.1\n",
       "ENSG00000239945  ENSG00000239945  Gene Expression   AL627309.3\n",
       "...                          ...              ...          ...\n",
       "ENSG00000277856  ENSG00000277856  Gene Expression   AC233755.2\n",
       "ENSG00000275063  ENSG00000275063  Gene Expression   AC233755.1\n",
       "ENSG00000271254  ENSG00000271254  Gene Expression   AC240274.1\n",
       "ENSG00000277475  ENSG00000277475  Gene Expression   AC213203.1\n",
       "ENSG00000268674  ENSG00000268674  Gene Expression      FAM231C\n",
       "\n",
       "[33538 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var['gene'] = adata.var.index\n",
    "adata.var.index = adata.var['gene_ids'].copy()\n",
    "adata.var_names = adata.var['gene_ids'].copy()\n",
    "adata.var.index.name = None\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "attached-awareness",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var.to_csv('/home/jovyan/farm/gene_ID_ENSG_table.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "suspended-pillow",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CTCL7    82203\n",
       "CTCL5    44882\n",
       "CTCL6    40151\n",
       "CTCL8    39126\n",
       "Name: donor_id, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['donor_id'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "another-horse",
   "metadata": {},
   "source": [
    "#### doublet removal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "isolated-staff",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_10x_channels = 'sample_id'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "billion-steam",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCL5_Derm_45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.35\n",
      "Detected doublet rate = 2.7%\n",
      "Estimated detectable doublet fraction = 28.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.3%\n",
      "Elapsed time: 22.9 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 5.4%\n",
      "Estimated detectable doublet fraction = 43.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 12.4%\n",
      "\n",
      "\n",
      "CTCL5_Derm_45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.1%\n",
      "Elapsed time: 12.8 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.1%\n",
      "Estimated detectable doublet fraction = 26.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.7%\n",
      "\n",
      "\n",
      "CTCL5_Derm_45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 0.9%\n",
      "Estimated detectable doublet fraction = 17.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.2%\n",
      "Elapsed time: 12.3 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 2.8%\n",
      "Estimated detectable doublet fraction = 31.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.9%\n",
      "\n",
      "\n",
      "CTCL5_Epi_45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.72\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 0.0%\n",
      "Elapsed time: 26.0 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.5%\n",
      "Estimated detectable doublet fraction = 21.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 16.3%\n",
      "\n",
      "\n",
      "CTCL5_Epi_45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.35\n",
      "Detected doublet rate = 1.5%\n",
      "Estimated detectable doublet fraction = 27.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.4%\n",
      "Elapsed time: 14.6 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.4%\n",
      "Estimated detectable doublet fraction = 43.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.7%\n",
      "\n",
      "\n",
      "CTCL5_Epi_45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 1.7%\n",
      "Estimated detectable doublet fraction = 22.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.8%\n",
      "Elapsed time: 21.2 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.8%\n",
      "Estimated detectable doublet fraction = 34.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "\n",
      "\n",
      "CTCL6_Derm45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 2.3%\n",
      "Estimated detectable doublet fraction = 26.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.8%\n",
      "Elapsed time: 19.8 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 6.0%\n",
      "Estimated detectable doublet fraction = 50.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.9%\n",
      "\n",
      "\n",
      "CTCL6_Derm45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.71\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 0.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.7%\n",
      "Elapsed time: 19.2 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.7%\n",
      "Estimated detectable doublet fraction = 27.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 13.4%\n",
      "\n",
      "\n",
      "CTCL6_Derm45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.76\n",
      "Detected doublet rate = 0.0%\n",
      "Estimated detectable doublet fraction = 0.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.0%\n",
      "Elapsed time: 16.3 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.6%\n",
      "Estimated detectable doublet fraction = 34.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.2%\n",
      "\n",
      "\n",
      "CTCL6_Epi45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 19.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.2%\n",
      "Elapsed time: 5.0 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 2.3%\n",
      "Estimated detectable doublet fraction = 31.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.2%\n",
      "\n",
      "\n",
      "CTCL6_Epi45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.5%\n",
      "Estimated detectable doublet fraction = 27.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 1.8%\n",
      "Elapsed time: 3.7 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 2.1%\n",
      "Estimated detectable doublet fraction = 45.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.6%\n",
      "\n",
      "\n",
      "CTCL6_Epi45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.30\n",
      "Detected doublet rate = 3.5%\n",
      "Estimated detectable doublet fraction = 34.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.2%\n",
      "Elapsed time: 18.6 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 5.1%\n",
      "Estimated detectable doublet fraction = 42.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.9%\n",
      "\n",
      "\n",
      "CTCL7_Derm45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.29\n",
      "Detected doublet rate = 4.8%\n",
      "Estimated detectable doublet fraction = 43.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "Elapsed time: 24.9 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 9.6%\n",
      "Estimated detectable doublet fraction = 63.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 15.1%\n",
      "\n",
      "\n",
      "CTCL7_Derm45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.19\n",
      "Detected doublet rate = 26.4%\n",
      "Estimated detectable doublet fraction = 68.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 38.7%\n",
      "Elapsed time: 74.6 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 0.2%\n",
      "Estimated detectable doublet fraction = 1.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 13.6%\n",
      "\n",
      "\n",
      "CTCL7_Derm45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.24\n",
      "Detected doublet rate = 11.2%\n",
      "Estimated detectable doublet fraction = 48.8%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 22.9%\n",
      "Elapsed time: 49.0 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 2.0%\n",
      "Estimated detectable doublet fraction = 13.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 14.5%\n",
      "\n",
      "\n",
      "CTCL7_Epi45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 0.8%\n",
      "Estimated detectable doublet fraction = 24.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 3.1%\n",
      "Elapsed time: 3.6 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 4.8%\n",
      "Estimated detectable doublet fraction = 46.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.1%\n",
      "\n",
      "\n",
      "CTCL7_Epi45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 1.8%\n",
      "Estimated detectable doublet fraction = 23.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.5%\n",
      "Elapsed time: 8.7 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 2.9%\n",
      "Estimated detectable doublet fraction = 33.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.6%\n",
      "\n",
      "\n",
      "CTCL7_Epi45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.37\n",
      "Detected doublet rate = 1.3%\n",
      "Estimated detectable doublet fraction = 23.3%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.7%\n",
      "Elapsed time: 4.6 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 4.5%\n",
      "Estimated detectable doublet fraction = 39.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.4%\n",
      "\n",
      "\n",
      "CTCL8_Derm45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.36\n",
      "Detected doublet rate = 3.4%\n",
      "Estimated detectable doublet fraction = 32.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 10.4%\n",
      "Elapsed time: 17.0 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 6.9%\n",
      "Estimated detectable doublet fraction = 53.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 13.0%\n",
      "\n",
      "\n",
      "CTCL8_Derm45P_8N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.36\n",
      "Detected doublet rate = 3.0%\n",
      "Estimated detectable doublet fraction = 33.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.2%\n",
      "Elapsed time: 15.6 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 6.9%\n",
      "Estimated detectable doublet fraction = 52.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 13.1%\n",
      "\n",
      "\n",
      "CTCL8_Derm45P_8Pr_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.52\n",
      "Detected doublet rate = 0.4%\n",
      "Estimated detectable doublet fraction = 7.7%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.5%\n",
      "Elapsed time: 17.7 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.6%\n",
      "Estimated detectable doublet fraction = 32.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 11.1%\n",
      "\n",
      "\n",
      "CTCL8_Epi45N_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 1.4%\n",
      "Estimated detectable doublet fraction = 22.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 6.4%\n",
      "Elapsed time: 10.2 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 3.3%\n",
      "Estimated detectable doublet fraction = 35.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 9.3%\n",
      "\n",
      "\n",
      "CTCL8_Epi45P_POOL_G\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.42\n",
      "Detected doublet rate = 1.6%\n",
      "Estimated detectable doublet fraction = 31.4%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.0%\n",
      "Elapsed time: 7.7 seconds\n",
      "Threshold found by scrublet\n",
      "Detected doublet rate = 4.4%\n",
      "Estimated detectable doublet fraction = 49.6%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.8%\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "RUNs, DSs, CELLs, THRs, MEDs, MADs, CUTs, no_thr = [], [], [], [], [], [], [], []\n",
    "\n",
    "for run in adata.obs[meta_10x_channels].unique():\n",
    "    print(run)\n",
    "    ad = adata[adata.obs[meta_10x_channels] == run, :]\n",
    "    x = ad.X\n",
    "    scrub = scr.Scrublet(x)\n",
    "    ds, prd = scrub.scrub_doublets()\n",
    "    RUNs.append(run)\n",
    "    DSs.append(ds)\n",
    "    CELLs.append(ad.obs_names)\n",
    "    # MAD calculation of threshold:\n",
    "    MED = np.median(ds)\n",
    "    MAD = robust.mad(ds)\n",
    "    CUT = (MED + (4 * MAD))\n",
    "    MEDs.append(MED)\n",
    "    MADs.append(MAD)\n",
    "    CUTs.append(CUT)\n",
    "\n",
    "    try:  # not always can calculate automatic threshold\n",
    "        THRs.append(scrub.threshold_)\n",
    "        print('Threshold found by scrublet')\n",
    "    except:\n",
    "        THRs.append(0.4)\n",
    "        no_thr.append(run)\n",
    "        print('No threshold found, assigning 0.4 to', run)\n",
    "        scrub.call_doublets(threshold=0.4) # so that it can make the plot\n",
    "    fig = scrub.plot_histogram()\n",
    "    fig[0].savefig(run + '.png')\n",
    "\n",
    "    # Alternative histogram for MAD-based cutoff\n",
    "    scrub.call_doublets(threshold=CUT)\n",
    "    fig = scrub.plot_histogram()\n",
    "    fig[0].savefig(run + '_mad_' + '.png')\n",
    "    plt.close('all')\n",
    "    print()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "scheduled-clearing",
   "metadata": {},
   "outputs": [],
   "source": [
    "ns = np.array(list(map(len, DSs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "incident-sphere",
   "metadata": {},
   "outputs": [],
   "source": [
    "tbl = pd.DataFrame({\n",
    "    'run': np.repeat(RUNs, ns),\n",
    "    'ds': np.concatenate(DSs),\n",
    "    'thr': np.repeat(THRs, ns),\n",
    "    'mad_MED': np.repeat(MEDs, ns),\n",
    "    'mad_MAD': np.repeat(MADs, ns),\n",
    "    'mad_thr': np.repeat(CUTs, ns),\n",
    "    }, index=np.concatenate(CELLs))\n",
    "\n",
    "tbl['auto_prd'] = tbl['ds'] > tbl['thr']\n",
    "tbl['mad_prd'] = tbl['ds'] > tbl['mad_thr']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "theoretical-samuel",
   "metadata": {},
   "outputs": [],
   "source": [
    "tbl.to_csv('/lustre/scratch117/cellgen/team205/rl20/CTCL/object-new/CTCL5678_doublets_score_mad.csv', header=True, index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "naked-concentration",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['mad_prd']=tbl['mad_prd']\n",
    "adata.obs['ds']=tbl['ds']\n",
    "adata.obs['mad_MED']=tbl['mad_MED']\n",
    "adata.obs['mad_MAD']=tbl['mad_MAD']\n",
    "adata.obs['mad_thr']=tbl['mad_thr']\n",
    "adata = adata[tbl['mad_prd'] != True]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "electrical-revelation",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "mito_genes = adata.var_names.str.startswith('MT-')\n",
    "adata.obs['percent_mito'] = np.sum(adata[:, mito_genes].X, axis=1).A1 / np.sum(adata.X, axis=1).A1\n",
    "adata.obs['n_counts'] = adata.X.sum(axis=1).A1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "loaded-cooler",
   "metadata": {},
   "outputs": [],
   "source": [
    "del(adata.obs['percent_mito'])\n",
    "del(adata.obs['n_counts'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "connected-specialist",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>mad_prd</th>\n",
       "      <th>ds</th>\n",
       "      <th>mad_MED</th>\n",
       "      <th>mad_MAD</th>\n",
       "      <th>mad_thr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTGATAA-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.129032</td>\n",
       "      <td>0.059282</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.205175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGTCATCCA-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.107930</td>\n",
       "      <td>0.059282</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.205175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAAGCCGTC-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.093656</td>\n",
       "      <td>0.059282</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.205175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCACCGTTGG-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>0.059282</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.205175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTTGCCTCT-1-18</th>\n",
       "      <td>CTCL5_Derm_45N_G</td>\n",
       "      <td>CTCL5</td>\n",
       "      <td>Dermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.038610</td>\n",
       "      <td>0.059282</td>\n",
       "      <td>0.036473</td>\n",
       "      <td>0.205175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCAACGCTA-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.094675</td>\n",
       "      <td>0.061414</td>\n",
       "      <td>0.033348</td>\n",
       "      <td>0.194805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCACGCGGT-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.061414</td>\n",
       "      <td>0.061414</td>\n",
       "      <td>0.033348</td>\n",
       "      <td>0.194805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCCAGATCA-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.048705</td>\n",
       "      <td>0.061414</td>\n",
       "      <td>0.033348</td>\n",
       "      <td>0.194805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCGGTCTAA-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.032599</td>\n",
       "      <td>0.061414</td>\n",
       "      <td>0.033348</td>\n",
       "      <td>0.194805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTTGTATC-1-40</th>\n",
       "      <td>CTCL8_Epi45P_POOL_G</td>\n",
       "      <td>CTCL8</td>\n",
       "      <td>Epidermis</td>\n",
       "      <td>False</td>\n",
       "      <td>0.041626</td>\n",
       "      <td>0.061414</td>\n",
       "      <td>0.033348</td>\n",
       "      <td>0.194805</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>198642 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 sample_id donor_id     Tissue  mad_prd  \\\n",
       "AAACCTGAGCTGATAA-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis    False   \n",
       "AAACCTGAGTCATCCA-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis    False   \n",
       "AAACCTGCAAGCCGTC-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis    False   \n",
       "AAACCTGCACCGTTGG-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis    False   \n",
       "AAACCTGGTTGCCTCT-1-18     CTCL5_Derm_45N_G    CTCL5     Dermis    False   \n",
       "...                                    ...      ...        ...      ...   \n",
       "TTTGTCATCAACGCTA-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis    False   \n",
       "TTTGTCATCACGCGGT-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis    False   \n",
       "TTTGTCATCCAGATCA-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis    False   \n",
       "TTTGTCATCGGTCTAA-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis    False   \n",
       "TTTGTCATCTTGTATC-1-40  CTCL8_Epi45P_POOL_G    CTCL8  Epidermis    False   \n",
       "\n",
       "                             ds   mad_MED   mad_MAD   mad_thr  \n",
       "AAACCTGAGCTGATAA-1-18  0.129032  0.059282  0.036473  0.205175  \n",
       "AAACCTGAGTCATCCA-1-18  0.107930  0.059282  0.036473  0.205175  \n",
       "AAACCTGCAAGCCGTC-1-18  0.093656  0.059282  0.036473  0.205175  \n",
       "AAACCTGCACCGTTGG-1-18  0.062500  0.059282  0.036473  0.205175  \n",
       "AAACCTGGTTGCCTCT-1-18  0.038610  0.059282  0.036473  0.205175  \n",
       "...                         ...       ...       ...       ...  \n",
       "TTTGTCATCAACGCTA-1-40  0.094675  0.061414  0.033348  0.194805  \n",
       "TTTGTCATCACGCGGT-1-40  0.061414  0.061414  0.033348  0.194805  \n",
       "TTTGTCATCCAGATCA-1-40  0.048705  0.061414  0.033348  0.194805  \n",
       "TTTGTCATCGGTCTAA-1-40  0.032599  0.061414  0.033348  0.194805  \n",
       "TTTGTCATCTTGTATC-1-40  0.041626  0.061414  0.033348  0.194805  \n",
       "\n",
       "[198642 rows x 8 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "satellite-square",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_file = '/lustre/scratch117/cellgen/team205/rl20/CTCL/object-new/CTCL5678_raw_rmdou.h5ad'\n",
    "adata.write(save_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "determined-bookmark",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (MyEnv_old)",
   "language": "python",
   "name": "myenv_old"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
